File: Assessment Time/Objective-C/LogViewController.m
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
The log view controller.
*/

#import "LogViewController.h"

@interface LogViewController ()
@property (weak) IBOutlet NSTextView *logTextView;
@property (strong) NSDateFormatter *formatter;
@end

@implementation LogViewController

id observer;

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Set up the formatter.
    self.formatter = [[NSDateFormatter alloc] init];
    [self.formatter setDateStyle:NSDateFormatterNoStyle];
    [self.formatter setTimeStyle:NSDateFormatterMediumStyle];
    
    // Set up the view.
    observer = [[NSNotificationCenter defaultCenter]
                addObserverForName:@"LogThis"
                object:nil
                queue:nil
                usingBlock:^(NSNotification * _Nonnull note) {
        NSString *message = [note object];
        if ( message ) {
            [self log:message];
        }
    }];
}

- (void)viewWillAppear {
    [super viewWillAppear];
    
    [[[self logTextView] window] setFrameAutosaveName:@"LogWindow"];
    [self log:@"Log view loaded."];
}

// Write the current date and a message to the log and to the console.
- (void) log:(NSString *)message {
    NSString *now = [self.formatter stringFromDate:[NSDate date]];
    NSString *messageWithDate = [NSString stringWithFormat:@"%@: %@\n",
                                 now, message];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.logTextView.textStorage appendAttributedString: [[NSAttributedString alloc] initWithString:messageWithDate]];
        [self.logTextView scrollToEndOfDocument:nil];
    });
    
    // Also log it to the console.
    NSLog(@"%@", message);
}
@end

File: Assessment Time/Swift/LogViewController.swift
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
The log view controller.
*/
import Cocoa

// Add the @objc attribute to be able to share a storyboard between Swift and Objective-C.
@objc(LogViewController)
class LogViewController: NSViewController {

    @IBOutlet var logTextView: NSTextView!
    var observer: Any?
    var formatter = DateFormatter()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Set up the formatter.
        formatter.dateStyle = .none
        formatter.timeStyle = .medium
        
        // Set up the view.
        observer = NotificationCenter.default.addObserver(forName: NSNotification.Name("LogThis"), object: nil, queue: nil) { note in
            if let message = note.object as? String {
                self.log(message)
            }
        }
    }
    override func viewWillAppear() {
        super.viewWillAppear()
        
        logTextView.window?.setFrameAutosaveName("LogWindow")
        log("Log view loaded.")
    }
    
    // Write the current date and a message to the log and to the console.
    func log(_ message: String ) {
        
        let now = formatter.string( from: Date() )

        let messageWithDate = "\(now): \(message)\n"
        DispatchQueue.main.async {
            self.logTextView.textStorage?.append( NSAttributedString(string: messageWithDate))
            self.logTextView.scrollToEndOfDocument(nil)
        }
        // Also log it to the console.
        NSLog("%@", message)
    }
}
